Opencascade loading of STEP files

stewartboogert opened the pull request:
* XCAF document loader

stewartboogert committed:
OCCCAD : XCAF based STEP file loader

New class to deal with step loading (G4OCCStepLoader). Based on XCAF. Simple example in load_step.cc

stewartboogert committed:
OCCCAD : Exploration of CAD file structure

G4OCCExplorer:
- label from name
- label from "entry"
- add topological hierarchy dump

Dump topology via TopDS_Explorer. Shape -> Face (surface) -> Wire -> Edge (curve)-> Vertex

Start of the extraction of surface parameters

stewartboogert committed:
OCCCAD : (cmake) update build to add libaries on all platforms.

stewartboogert committed:
OCCCAD : (tests) testing new OCC functionality

stewartboogert committed:
OCCCAD : (cmake) Fix for docker OpenCASCADE dependency

dboogert commented on CMakeLists.txt:
```diff
@@ -33,20 +33,31 @@ include(${Geant4_USE_FILE})
 find_package(VTK REQUIRED)
 find_package(Catch2 REQUIRED)
-
+find_package(OpenCASCADE  COMPONENTS FoundationClasses ModelingAlgorithms
```
I think you need to add `ApplicationFramework` & `Draw` here to get all the targets.

dboogert commented on include/occCAD/G4OCCExplorer.hh:
```diff
@@ -0,0 +1,38 @@
+// Created by Boogert, Stewart (-,DL,-) on 27/07/2023.
+//
+
+#ifndef GEANT4FX_G4OCCEXPLORER_HH
```
```suggestion
#pragma once
```

dboogert commented on include/occCAD/G4OCCStepLoader.hh:
```diff
@@ -0,0 +1,28 @@
+
+class G4OCCStepLoader {
+public:
+    G4OCCStepLoader(std::string file_name);
```
```suggestion
    G4OCCStepLoader(const std::string& file_name);
```

dboogert commented on include/occCAD/G4OCCExplorer.hh:
```diff
@@ -0,0 +1,38 @@
+    //G4OCCExplorer();
+    //~G4OCCExplorer();
+
+    static void Dump(Handle(XCAFDoc_ShapeTool) shapeTool);
```
What is the Handle macro? Is it different from opencascade::handle?

dboogert commented on include/occCAD/G4OCCExplorer.hh:
```diff
@@ -0,0 +1,38 @@
+    //~G4OCCExplorer();
+
+    static void Dump(Handle(XCAFDoc_ShapeTool) shapeTool);
+    static opencascade::handle<TDataStd_Name> GetNameFromLabel(TDF_Label &label);
```
Can the parameter and function be `const`? I would check all functions in G4OCCExplorer & G4OCCStepLoader

dboogert commented on src/test/load_step.cc:
```diff
@@ -0,0 +1,35 @@
+#include "occCAD/G4OCCExplorer.hh"
+
+int main() {
+    auto step_reader = G4OCCStepLoader("03_Cube_with_fillet.step");
```
How is this different from the test?

dboogert commented:
The execution of the tests should be added to the CI workflow.

stewartboogert commented:
> The execution of the tests should be added to the CI workflow.

Ok. Need to know how to edit the `build.yaml` for this. I have commited the changes to came and installing the step files. It really is just how to include in the workflow config

stewartboogert committed:
OCCCAD : (test) Install test code and example step file

dboogert commented on include/occCAD/G4OCCStepLoader.hh:
```diff
@@ -0,0 +1,28 @@
+//
```
I'd remove the auto generated header.

dboogert commented on src/occCAD/G4OCCExplorer.cc:
```diff
@@ -0,0 +1,190 @@
+    while(exp.More()) {
+        auto current = exp.Current();
+
+        //current.DumpJson(std::cout);
```
I'd remove commented out code - if you want to leave a note for yourself then write up some comments.

dboogert commented on src/occCAD/G4OCCExplorer.cc:
```diff
@@ -0,0 +1,190 @@
+}
+
+void G4OCCExplorer::DumpFace(TopoDS_Face &face) {
+    std::cout << "DumpFace" << std::endl;
```
All this Dump stuff you could write to a std::stream& and pass in a stringstream to verify in the tests. This might be over the top.

dboogert reviewed (approved):
LGTM.

Lots of nitpicking. 

commit `[OCCCAD : (cmake) Fix for docker OpenCASCADE dependency](https://github.com/g4edge/g4fx/pull/3/commits/5e95299098236005ace3deea5454cbac69898188)` can be fixed up into first commit you update the cmake file.

Last two commits can be merged into one.

stewartboogert committed:
OCCCAD : (review) Fixed review comments

- no comments of unused code
- dump takes std::ostream
- const where possible
- remove Handle(type) and changed to opencascade::handle<type>
